internal/cephfs/groupcontrollerserver.go

```diff
@@ -0,0 +1,758 @@
+
+		return err
+	}
+
+	for volID, snapID := range vgsi.VolumeSnapshotMap {
+		// delete the snapshots
+		req := &csi.DeleteSnapshotRequest{
+			SnapshotId: snapID,
+			Secrets:    secrets,
+		}
+		_, err = cs.DeleteSnapshot(ctx, req)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to delete snapshot: %v", err)
+
+			return err
+		}
+
+		j, err := store.VolumeGroupJournal.Connect(bgi.Monitors, fsutil.RadosNamespace, cr)
+		if err != nil {
+			return err
+		}
+		// remove the entry from the omap
+		err = j.RemoveVolumeSnapshotMapping(
+			ctx,
+			bgi.MetadataPool,
+			vgsi.ReservedID,
+			volID)
+		j.Destroy()
+		if err != nil {
+			log.ErrorLog(ctx, "failed to remove volume snapshot mapping: %v", err)
+
+			return err
+		}
+		// undo the reservation
+		err = store.UndoVolumeGroupReservation(ctx, bgi, vgsi, cr)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to undo volume group reservation: %v", err)
+
+			return err
+		}
+	}
+
+	for _, fm := range fsMap {
+		_, err := fm.ResetFSQuiesce(ctx, vgs.RequestName)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to reset filesystem quiesce: %v", err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+// validateVolumeGroupSnapshotDeleteRequest validates the request for creating a group
+// snapshot of volumes.
+func (cs *ControllerServer) validateVolumeGroupSnapshotDeleteRequest(
+	ctx context.Context,
+	req *csi.DeleteVolumeGroupSnapshotRequest,
+) error {
+	if err := cs.Driver.ValidateGroupControllerServiceRequest(
+		csi.GroupControllerServiceCapability_RPC_CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT); err != nil {
+		log.ErrorLog(ctx, "invalid create volume group snapshot req: %v", protosanitizer.StripSecrets(req))
+
+		return err
+	}
+
+	// Check sanity of request volume group snapshot Name, Source Volume Id's
+	if req.GetGroupSnapshotId() == "" {
+		return status.Error(codes.InvalidArgument, "volume group snapshot id cannot be empty")
+	}
+
+	return nil
+}
+
+// DeleteVolumeGroupSnapshot deletes a group snapshot of volumes.
+func (cs *ControllerServer) DeleteVolumeGroupSnapshot(ctx context.Context,
+	req *csi.DeleteVolumeGroupSnapshotRequest) (
+	*csi.DeleteVolumeGroupSnapshotResponse,
+	error,
+) {
+	if err := cs.validateVolumeGroupSnapshotDeleteRequest(ctx, req); err != nil {
+		return nil, err
+	}
+
+	groupSnapshotID := req.GroupSnapshotId
+	// Existence and conflict checks
+	if acquired := cs.VolumeGroupLocks.TryAcquire(groupSnapshotID); !acquired {
+		log.ErrorLog(ctx, util.VolumeOperationAlreadyExistsFmt, groupSnapshotID)
+
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, groupSnapshotID)
+	}
+	defer cs.VolumeGroupLocks.Release(groupSnapshotID)
+
+	cr, err := util.NewAdminCredentials(req.GetSecrets())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	defer cr.DeleteCredentials()
+
+	_, vgsi, err := store.NewVolumeGroupOptionsFromID(ctx, req.GroupSnapshotId, cr)
```

Comment:
The `_` variable needs a `Destroy()` here too, I think?

Reply:
yes its a good catch, fixed it now